### ISSUE TITLE
Conflict with std

### ISSUE THREAD

**Issue by weslleyspereira** [tlapack/tlapack#134]
@alilofti90 is implementing LU with partial pivoting. This routine needs to swap rows of the matrix. One of his issues was that the call `swap( row1, row2 )` actually calls `std::swap` and not `tlapack::swap`, even when we are inside the namespace `tlapack`. This is the same problem of using `tlapack::abs`.

- We should avoid name conflicts with std. I suggest we rename `swap` to `deep_swap`, `swap_arrays` or similar.

- It is known `std::complex<T>::abs` may not propagate NaNs. This is the only reason \<T\>LAPACK has another implementation. This is an issue to be fixed upstream, i.e., in the std library. So (1) if we use std::complex<T> as the complex type, we live with the NaN propagation issue; (2) we may specialize `std::complex` for our routines, so that we can redefine `std::abs`. I would need to test that (2) is a valid option.

### Discussed in https://github.com/tlapack/tlapack/discussions/133

<div type='discussions-op-text'>

<sup>Originally posted by **alilotfi90** September  6, 2022</sup>
Hi,
There is a conflict between std swap and tlapack swap. 
</div>

**Comment by langou** [comment]
"We should avoid name conflicts with std. I suggest we rename swap to deep_swap, swap_arrays or similar." (*) We can always do tlapack::sway() to make sure we are calling what we think we are calling. (*) Avoiding conflicts with std is fine, but isn't it why we can have the tlapack:: at the start? So that if there is a name conflict, this is resolved. (*) Since no one seems to be keen on using tlapack:: all the time, then my preference is to rename tlapack_swap then. I think just be explicit with the name conflict. I know that tlapack::tlapack_swap is ugly, but it represents where we are. Hiding the issue with a name such as deep_swap or swap_arrays is not a good idea in my opinion. Just be explicit about it. tlapack::tlapack_swap

**Comment by weslleyspereira** [comment]
I understand your point. We could just always use `tlapack::`. This is expected if we are writing tests or code using the library. When we are implementing routines inside the namespace `tlapack`, we can avoid wordiness by suppressing `tlapack::`.

I have just tried to reproduce my issue with `swap` on a simple example, but I couldn't. So, I think `tlapack::swap` is in good shape.

I want to detail the case of `std::abs`:

```c++
#include <complex>

namespace tlapack {

    template< typename T >
    inline auto abs( const T& x ) {
        // Do stuff
    }

    int getf0() {

        using tlapack::abs;
        
        float t;
        abs( t ); // calls tlapack::abs.
        
        long double u;
        abs( u ); // calls tlapack::abs.
        
        std::complex<float> v;
        abs( v ); // calls std::abs.
        
        std::complex<long double> w;
        abs( w ); // calls std::abs.
        
        std::complex<long double> x;
        tlapack::abs( x ); // calls tlapack::abs.

        return 0;
    }
}

int main() {

    using tlapack::abs;
    
    float t;
    abs( t ); // calls tlapack::abs.
    
    long double u;
    abs( u ); // calls tlapack::abs.
    
    std::complex<float> v;
    abs( v ); // calls std::abs.
    
    std::complex<long double> w;
    abs( w ); // calls std::abs.
    
    std::complex<long double> x;
    tlapack::abs( x ); // calls tlapack::abs.

    tlapack::getf0();

    return 0;
}
```

This is because of the Argument-Dependent Lookup (ADL). Since the call `abs` is unqualified, the compiler will look at the namespace of the argument `std::vector`, so namespace `std`, and then find the function `std::abs` there. Even thought the we say `using tlapack::abs;`, the unqualified call will refer to `std::abs`. For the same reason, we cannot define:

```c++
namespace tlapack {
    template< typename T >
    inline T abs( const std::vector<T>& x ) {
        // Do stuff
    }
}
```

Complete explanation here: https://en.cppreference.com/w/cpp/language/adl .

**Comment by weslleyspereira** [comment]
Hi,

I solved the problem of `std::abs` in #145 by mimicking the interface from https://en.cppreference.com/w/cpp/numeric/complex/abs inside the namespace `tlapack`. So, now we enforce (at least for the GCC compiler) that one should use either `std::abs` or `tlapack::abs` when the input is `std::complex<T>`.

#146 solves the issue related to `swap`. Again, this is related to ADL and `std::complex`.
